activerecord/test/cases/associations_test.rb

```diff
@@ -147,6 +147,15 @@ def test_belongs_to_a_model_with_composite_primary_key_uses_composite_pk_in_sql
     assert_match(/#{Regexp.escape(Sharded::BlogPost.connection.quote_table_name("sharded_blog_posts.id"))} =/, sql)
   end
 
+  def test_querying_by_whole_associated_records_using_query_constraints
```

Comment:
Let me know if you think there is a better place to put this test in. Since I don't think we have a common name for the `association_name: records` querying capability I struggled to find similar tests to see if the new test could live together with tests for regular models. So I decided to put it in `associations_test` to live close to most of our `query_constraints` in associations test.